src/org/sipdroid/sipua/PublishAgent.java

```diff
@@ -0,0 +1,222 @@
+			Message req = msgf.createPublishRequest(sip_provider, new NameAddress(from), "presence", expireTime, "application/pidf+xml", xml);
+			TransactionClient t = new TransactionClient(sip_provider, req, this);
+			t.request();
+		}
+	}
+
+	public void unPublish() {
+		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+		String from = user_profile.username;
+		Boolean publish_enable_status = prefs.getBoolean("publish_enable", true);
+		if (publish_enable_status == true) {
+			MessageFactory msgf = new MessageFactory();
+			Message req = msgf.createPublishRequest(sip_provider, new NameAddress(from), "presence", 0, null, null);
+			TransactionClient t = new TransactionClient(sip_provider, req, this);
+			t.request();
+		}
+	}
+
+	private boolean generateRequestWithProxyAuthorizationheader(Message resp, Message req) {
+		ProxyAuthenticateHeader pah = resp.getProxyAuthenticateHeader();
+		ProxyAuthorizationHeader ah1 = (new DigestAuthentication(
+				req.getTransactionMethod(), req.getRequestLine().getAddress()
+				.toString(), pah, null, null, username, passwd))
+				.getProxyAuthorizationHeader();
+		req.setProxyAuthorizationHeader(ah1);
+		return true;
+	}
+
+	public void onTransSuccessResponse(TransactionClient tc, Message resp) {
+	}
+
+	public void onTransFailureResponse(TransactionClient tc, Message resp) {
+		StatusLine status = resp.getStatusLine();
+		System.out.println("CheckCheck");
+		int code = status.getCode();
+		processAuthenticationResponse(tc, resp, code);
+	}
+
+	private boolean processAuthenticationResponse(TransactionClient transaction, Message resp, int respCode) {
+		Message req = transaction.getRequestMessage();
+		req.setCSeqHeader(req.getCSeqHeader().incSequenceNumber());
+		ViaHeader vh = req.getViaHeader();
+		String newbranch = SipProvider.pickBranch();
+		vh.setBranch(newbranch);
+		req.removeViaHeader();
+		req.addViaHeader(vh);
+		handleAuthentication(respCode, resp, req);
+		TransactionClient t = new TransactionClient(sip_provider, req, this, 30000);
+		t.request();
+		return true;
+	}
+
+	private boolean generateRequestWithWwwAuthorizationheader(Message resp, Message req) {
+		WwwAuthenticateHeader wah = resp.getWwwAuthenticateHeader();
+		AuthorizationHeader ah = (new DigestAuthentication(
+				req.getTransactionMethod(), req.getRequestLine().getAddress()
+				.toString(), wah, null, null, user_profile.username, user_profile.passwd))
+				.getAuthorizationHeader();
+		req.setAuthorizationHeader(ah);
+		return true;
+
+	}
+
+	private boolean handleAuthentication(int respCode, Message resp, Message req) {
+		switch (respCode) {
+			case 407:
+				return generateRequestWithProxyAuthorizationheader(resp, req);
+			case 401:
+				return generateRequestWithWwwAuthorizationheader(resp, req);
+		}
+		return false;
+	}
+
+	@Override
+	public void onTransTimeout(TransactionClient tc) {
+
+	}
+
+	public void onTransProvisionalResponse(TransactionClient tc, Message resp) {
+

```

Comment:
put a log statement here, log the response code and status line

include a FIXME comment for this method to be fully implemented